### PR TITLE
api: resource.k8s.io v1alpha1 -> v1alpha2

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

For Kubernetes 1.27, we intend to make some breaking API changes:
- rename PodScheduling -> PodSchedulingHints (https://github.com/kubernetes/kubernetes/issues/114283)
- extend ResourceClaimStatus (https://github.com/kubernetes/enhancements/pull/3802)

We need to switch from v1alpha1 to v1alpha2 for that.

#### Special notes for your reviewer:

The actual API changes will follow in separate PRs.

#### Does this PR introduce a user-facing change?
```release-note
ACTION REQUIRED: resource.k8s.io/v1alpha1 was replaced with resource.k8s.io/v1alpha2. Before upgrading a cluster, all objects in resource.k8s.io/v1alpha1 (ResourceClaim, ResourceClaimTemplate, ResourceClass, PodScheduling) must be deleted. The changes will be internal, so YAML files which create pods and resource claims don't need changes except for the newer `apiVersion`.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3063
```

### CODE REVIEW

**Comment by klueska** [comment]
From what I can tell looking at other APIs that have bumped their versions, you actually want to create a new `pkg/apis/resource/v1alpha1` folder as part of this PR, not a `pkg/apis/resource/v1alpha2` (or optionally both, but not just `v1alpha2`).

It is then the job of the code in the `v1alpha1` package to implement code that knows how to convert to the latest type up in `pkg/apis/resource` (not the other way around).

**Comment by pohly** [comment]
> From what I can tell looking at other APIs that have bumped their versions

Do you have an example? There is currently no other v1alpha2 API in the tree.

> you actually want to create a new pkg/apis/resource/v1alpha1 folder as part of this PR

Create? pkg/apis/resource/v1alpha1 is what we have right now. We would have to *keep* it and *add* v1alpha2. That is the approach that is taken when two different versions need to be supported concurrently. But we want to drop support for v1alpha1 and replace it with v1alpha2, to avoid writing conversion code.

**Comment by klueska** [comment]
OK, that makes sense, and is consistent with what we talked about in person yesterday. Could you then tell me how you invoke `controller-gen` to generate the `v1alpha2` folder and its contents? I can't change anything in `pkg/apis/resource` without regenerating this since it then complains I don't have proper conversion functions between `pkg/apis/resource/v1alpha2` and `pkg/apis/resource`.

**Comment by pohly** [comment]
/unassign @thockin 

I'd like to get an LGTM first from someone in SIG Node first (as confirmation that this change is desired), then we can pick an API approver. This is mostly a mechanical change, so no familiarity with the feature is needed and thus it doesn't need to be reviewed by Tim.

**Comment by klueska** [comment]
As long as this passes the API review for bumping to `v1alpha2` without a migration path from `v1alpha1` then this PR looks good to me. The explicit goal was to replace `v1alpha1` with `v1alpha2` without supporting migration from one to another and the PR does exactly that.

/lgtm
/approve

**Comment by liggitt** [comment]
indicate in the release note anyone using v1alpha1 must delete all existing object instances before upgrading their server to 1.27

File: cmd/kube-apiserver/app/aggregator.go [hunk 1]
```diff
@@ -283,7 +283,7 @@ var apiVersionPriorities = map[schema.GroupVersion]priority{
 	{Group: "flowcontrol.apiserver.k8s.io", Version: "v1beta1"}:  {group: 16100, version: 12},
 	{Group: "flowcontrol.apiserver.k8s.io", Version: "v1alpha1"}: {group: 16100, version: 9},
 	{Group: "internal.apiserver.k8s.io", Version: "v1alpha1"}:    {group: 16000, version: 9},
-	{Group: "resource.k8s.io", Version: "v1alpha1"}:              {group: 15900, version: 9},
+	{Group: "resource.k8s.io", Version: "v1alpha2"}:              {group: 15900, version: 9},
```
Comment (thockin) [inline cmd/kube-apiserver/app/aggregator.go:286]: Don't yhou want to keep both of these?  Or is alpha1 just nuked totally?

File: hack/lib/init.sh [hunk 2]
```diff
@@ -92,7 +92,7 @@ coordination.k8s.io/v1beta1 \
 coordination.k8s.io/v1 \
 discovery.k8s.io/v1 \
 discovery.k8s.io/v1beta1 \
-resource.k8s.io/v1alpha1 \
```
Comment (thockin) [inline hack/lib/init.sh:95]: Don't you want both?

File: pkg/api/testing/defaulting_test.go [hunk 3]
```diff
@@ -137,10 +137,10 @@ func TestDefaulting(t *testing.T) {
 		{Group: "rbac.authorization.k8s.io", Version: "v1", Kind: "ClusterRoleBindingList"}:                        {},
 		{Group: "rbac.authorization.k8s.io", Version: "v1", Kind: "RoleBinding"}:                                   {},
 		{Group: "rbac.authorization.k8s.io", Version: "v1", Kind: "RoleBindingList"}:                               {},
-		{Group: "resource.k8s.io", Version: "v1alpha1", Kind: "ResourceClaim"}:                                     {},
```
Comment (thockin) [inline pkg/api/testing/defaulting_test.go:140]: Last time I ask - don't you want both?

**Review by thockin — COMMENTED** [review]
Totally obliterating an API, even alpha, at the same time as replacing it is unusual.  It means anyone who has a saved object of this type will be unable to delete.  That's pretty dramatic.  Is it warranted?  

If you keep both you will have to manually write conversion for things that you change in alpha2, but that's easy enough.

**Comment by pohly** [comment]
> If you keep both you will have to manually write conversion for things that you change in alpha2, but that's easy enough.

Is it really easy when completely renaming a type? I know how to handle conversion of individual fields. I'm not sure what to do for the `PodScheduling -> PodSchedulingHints` renaming. Convert between the types? Treat them as separate?

What @klueska plans to do in https://github.com/kubernetes/kubernetes/pull/116332 also won't be easy: there's no good mapping between v1alpha1 
```
type AllocationResult struct {
    ResourceHandle string
    AvailableOnNodes *core.NodeSelector
    Shareable bool
}
```

and v1alpha2
```
type AllocationResult struct {
    NodePluginData []ResourceHandle 
    AvailableOnNodes *core.NodeSelector
    Shareable bool
}

type ResourceHandle struct {
    PluginName string
    Data string
}
```

Conversion from v1alpha1 to v1alpha2 works, but the other direction is lossy.

I would prefer to nuke v1alpha1 instead of spending time and effort into writing code which, in practice, will not be used much. All of the DRA drivers are experimental. I doubt that there are many people who will upgrade from a 1.26 cluster with DRA.

**Comment by pohly** [comment]
This is probably the last time we can do something like this. v1alpha2 will be more widely used.

**Comment by pohly** [comment]
I discussed it with @klueska and we agreed to try the approach with both v1alpha1 and v1alpha2 supported concurrently.

I created a [separate PR](https://github.com/kubernetes/kubernetes/pull/116458) with that approach. If that works out, we can merge that instead of this one here.

**Comment by thockin** [comment]
Yeah, I don't think we know how to rename a top-level type across API versions, but I don't know if we have ever tried.

**Comment by thockin** [comment]
/approve
/lgtm

/hold

(while we figure out which path)

**Comment by thockin** [comment]
This relnote will need editing after type renaming - it seems to say "no change needed except version" :)

/remove-hold

**Comment by thockin** [comment]
For the record, we discussed and decided the cost of keeping v1alpha1 was not justified, and the breakage of simply removing it is low (and alpha).

**Comment by pohly** [comment]
Yep, the release notes are for this PR only. The follow-up PRs will have additional instructions. I've been told the release team will merge related entries.

**Comment by pohly** [comment]
Rebased because of a merge conflict in `api/openapi-spec`. "git rebase" resolved that automatically?

**Comment by klueska** [comment]
/lgtm
after rebase

**Pull request merged**